test/Driver/single_stack_test.jl

```diff
@@ -0,0 +1,615 @@
+
+# ## Inspect the initial conditions for the first node
+
+# Let's export a plot of the initial state
+output_dir = @__DIR__;
+
+mkpath(output_dir);
+
+z_scale = 100 # convert from meters to cm
+z_key = "z"
+z_label = "z [cm]"
+z = get_z(driver_config.grid, z_scale)
+state_vars = get_vars_from_nodal_stack(
+    driver_config.grid,
+    solver_config.Q,
+    vars_state_conservative(m, FT),
+    i = 1,
+    j = 1,
+);
+aux_vars = get_vars_from_nodal_stack(
+    driver_config.grid,
+    solver_config.dg.state_auxiliary,
+    vars_state_auxiliary(m, FT),
+    i = 1,
+    j = 1,
+    exclude = [z_key],
+);
+all_vars = OrderedDict(state_vars..., aux_vars...);
+
+# Generate plots of initial conditions
+export_plot_snapshot(
+    z,
+    all_vars,
+    ("ρcT",),
+    joinpath(output_dir, "initial_condition_T.png"),
+    z_label,
+);
+export_plot_snapshot(
+    z,
+    all_vars,
+    ("ρu[1]",),
+    joinpath(output_dir, "initial_condition_u.png"),
+    z_label,
+);
+export_plot_snapshot(
+    z,
+    all_vars,
+    ("ρu[2]",),
+    joinpath(output_dir, "initial_condition_v.png"),
+    z_label,
+);
+export_plot_snapshot(
+    z,
+    all_vars,
+    ("ρu[3]",),
+    joinpath(output_dir, "initial_condition_w.png"),
+    z_label,
+);
+
+# Horizontal statistics of variables
+state_vars_var = get_horizontal_variance(
+    driver_config.grid,
+    solver_config.Q,
+    vars_state_conservative(m, FT),
+);
+
+state_vars_avg = get_horizontal_mean(
+    driver_config.grid,
+    solver_config.Q,
+    vars_state_conservative(m, FT),
+);
+
+export_plot_snapshot(
+    z,
+    state_vars_avg,
+    ("ρu[1]",),
+    joinpath(output_dir, "initial_condition_avg.png"),
+    z_label,
+);
+export_plot_snapshot(
+    z,
+    state_vars_var,
+    ("ρu[1]",),
+    joinpath(output_dir, "initial_condition_variance.png"),
+    z_label,
+);
+
+# ![](initial_condition_avg.png)
+# ![](initial_condition_variance.png)
+
+# # Solver hooks / callbacks
+
+# Define the number of outputs from `t0` to `timeend`
+const n_outputs = 5;
+
+# This equates to exports every ceil(Int, timeend/n_outputs) time-step:
+const every_x_simulation_time = ceil(Int, timeend / n_outputs);
+
+# Create a dictionary for `z` coordinate (and convert to cm) NCDatasets IO:
+dims = OrderedDict(z_key => collect(z));
+
+data_var = Dict([k => Dict() for k in 0:n_outputs]...)
+data_var[0] = deepcopy(state_vars_var)
+
+data_avg = Dict([k => Dict() for k in 0:n_outputs]...)
+data_avg[0] = deepcopy(state_vars_avg)
+# The `ClimateMachine`'s time-steppers provide hooks, or callbacks, which
+# allow users to inject code to be executed at specified intervals. In this
+# callback, the state and aux variables are collected, combined into a single
+# `OrderedDict` and written to a NetCDF file (for each output step `step`).
+step = [0];
+callback = GenericCallbacks.EveryXSimulationTime(
+    every_x_simulation_time,
+    solver_config.solver,
+) do (init = false)
+    state_vars_var = get_horizontal_variance(
```

Comment:
Just to be clear, this is not computing averages at every timestep. For that behavior, you'd want to put `get_horizontal_variance` in `update_auxiliary_state!`. In other words, right now, your solution depends on the frequency of your callbacks.

Reply:
I know. I just implemented functions that provide the horizontal statistics diagnostics for outputs. I did not want to compute them for every timestep since I am not using the horizontal statistics for the evolution of the system.